test/jpypetest/test_leak2.py

```diff
@@ -0,0 +1,80 @@
+import _jpype
+import jpype
+from jpype.types import *
+import sys
+import logging
+import time
+import common
+
+class Tracer(object):
+    ctor = 0
+    dtor = 0
+    def __init__(self):
+        Tracer.ctor +=1 
+    def __del__(self):
+        Tracer.dtor +=1 
+
+    @staticmethod
+    def reset():
+        Tracer.ctor = 0
+        Tracer.dtor = 0
+
+    @staticmethod
+    def leaked():
+        return Tracer.ctor-Tracer.dtor
+
+    @staticmethod
+    def attach(obj):
+        object.__setattr__(obj, "_trace", Tracer())
+    
+
+# This test finds reference counting leak by attaching an 
+# object that lives as long as the wrapper is alive.  
+# It can't detect Java reference counter leaks.
+class Leak2TestCase(common.JPypeTestCase):
+
+    def setUp(self):
+        common.JPypeTestCase.setUp(self)
+        self.fixture = JClass('jpype.common.Fixture')()
+        Tracer.reset()
+
+    def testArrayCall(self):
+        JA = JArray(JInt)
+        def call():
+            inst = JA(100)
+            Tracer.attach(inst)
+            self.fixture.callObject(inst)
+        for i in range(100):
+            call()
+        self.assertEqual(Tracer.leaked(), 0)
```

Comment:
You have ensured that no intermediate layer has modified the refcount of the JPArray instance.

Reply:
Yeah.  This pattern was submitted by a user on the bug report for memory exhaustion.  By attaching a Python object to a wrapper it verifies that we didn't accidentally leak the Python portion of the wrapper.   It does check for one type of leak.  Leaks on the Java side remain difficult.  Although we can ask Java to run a garbage collection is it just a suggestion.  Thus writing tests to see if we leaked a Java object are more challenging as we can't guarantee that  the object is collected.

I have debated adding a atomic integer counter to JPContext (when compiled with instrumentation) so that we can actually get the current number of global references that we are holding.  We would then get the count before and after the operation.  But I haven't gotten around to it and like everything having to do with threads I worry about the portability of such a trick for code in general.   I got bitten once already by the JPTracer where  the function back referencing was causing a seg fault.  Obviously a counter for auditing is much less risk.

Reply:
keep it simple stupid :)